### PR TITLE
Issue 4596 - BUG - lto=thin

### LINKED ISSUE #4596: Build with clang/lld fails with undefined reference error when LTO is enabled
**Issue Description**
In Fedora 33 LTO is enabled by default: https://fedoraproject.org/wiki/LTOByDefault
Build with clang/lld fails with undefined reference error.

**Steps to Reproduce**
On Fedora 33 with 
```
lld-11.0.0-1.fc33.x86_64
clang-11.0.0-2.fc33.x86_64
llvm-libs-11.0.0-1.fc33.x86_64
```

```
$ sed -i -e 's/COCKPIT_ON = 1/COCKPIT_ON = 0/g' -e 's/CLANG_ON = 0/CLANG_ON = 1/g' rpm.mk
$ make -f rpm.mk rpms
```

```
/bin/sh ./libtool  --tag=CC   --mode=link clang -g -O2 -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1     -O2 -flto -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS --config /usr/lib/rpm/redhat/redhat-hardened-clang.cfg -fstack-protector-strong   -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -avoid-version -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now  -flto -o libattr-unique-plugin.la -rpath /usr/lib64/dirsrv/plugins ldap/servers/plugins/uiduniq/libattr_unique_plugin_la-7bit.lo ldap/servers/plugins/uiduniq/libattr_unique_plugin_la-uid.lo ldap/servers/plugins/uiduniq/libattr_unique_plugin_la-utils.lo libslapd.la -lplds4 -lplc4 -lnspr4 -lpthread -ldl  
ld.lld: error: ./.libs/libslapd.so: undefined reference to __rust_probestack [--no-allow-shlib-undefined]
ld.lld: error: ./.libs/libslapd.so: undefined reference to __muloti4 [--no-allow-shlib-undefined]
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

If I disable LTO in the spec file, build succeeds:
```
%global _lto_cflags %nil
```

### PR DESCRIPTION
**Pull request by Firstyear**
Bug Description: There were some issues building with lto thing + rust

Fix Description: Change our config.toml to support this.

fixes: https://github.com/389ds/389-ds-base/issues/4596

Author: William Brown <william@blackhats.net.au>

Review by: ???

### CODE REVIEW

**Comment by Firstyear** [comment]
@vashirov Can you try this? I had some issues however with getting clang++ to work with dshttpd, so let me know if you hit issues. I think this is really pushing us to the limits of what autotools can actually do though, so we really may need to consider a buildsystem replacement in the future.

**Comment by vashirov** [comment]
This doesn't work for me, linker fails with SIGABRT:
```
*** Warning: Linking the shared library libslapd.la against the
*** static library /home/vashirov/src/ds/rpmbuild/BUILD/389-ds-base-2.1.0.20210913git9eaf1f35e/rs/rslapd/release/librslapd.a is not portable!
libtool: link: clang -shared  -fPIC -DPIC  ldap/servers/slapd/.libs/libslapd_la-add.o ldap/servers/slapd/.libs/libslapd_la-agtmmap.o ldap/servers/slapd/.libs/libslapd_la-apibroker.o ldap/servers/slapd/.libs/libslapd_la-attr.o ldap/servers/slapd/.libs/libslapd_la-attrlist.o ldap/servers/slapd/.libs/libslapd_la-attrsyntax.o ldap/servers/slapd/.libs/libslapd_la-auditlog.o ldap/servers/slapd/.libs/libslapd_la-ava.o ldap/servers/slapd/.libs/libslapd_la-backend.o ldap/servers/slapd/.libs/libslapd_la-backend_manager.o ldap/servers/slapd/.libs/libslapd_la-bitset.o ldap/servers/slapd/.libs/libslapd_la-bulk_import.o ldap/servers/slapd/.libs/libslapd_la-charray.o ldap/servers/slapd/.libs/libslapd_la-ch_malloc.o ldap/servers/slapd/.libs/libslapd_la-computed.o ldap/servers/slapd/.libs/libslapd_la-control.o ldap/servers/slapd/.libs/libslapd_la-configdse.o ldap/servers/slapd/.libs/libslapd_la-counters.o ldap/servers/slapd/.libs/libslapd_la-csn.o ldap/servers/slapd/.libs/libslapd_la-csngen.o ldap/servers/slapd/.libs/libslapd_la-csnset.o ldap/servers/slapd/.libs/libslapd_la-defbackend.o ldap/servers/slapd/.libs/libslapd_la-delete.o ldap/servers/slapd/.libs/libslapd_la-dl.o ldap/servers/slapd/.libs/libslapd_la-dn.o ldap/servers/slapd/.libs/libslapd_la-dse.o ldap/servers/slapd/.libs/libslapd_la-dynalib.o ldap/servers/slapd/.libs/libslapd_la-entry.o ldap/servers/slapd/.libs/libslapd_la-entrywsi.o ldap/servers/slapd/.libs/libslapd_la-errormap.o ldap/servers/slapd/.libs/libslapd_la-eventq.o ldap/servers/slapd/.libs/libslapd_la-eventq-deprecated.o ldap/servers/slapd/.libs/libslapd_la-factory.o ldap/servers/slapd/.libs/libslapd_la-features.o ldap/servers/slapd/.libs/libslapd_la-fileio.o ldap/servers/slapd/.libs/libslapd_la-filter.o ldap/servers/slapd/.libs/libslapd_la-filtercmp.o ldap/servers/slapd/.libs/libslapd_la-filterentry.o ldap/servers/slapd/.libs/libslapd_la-generation.o ldap/servers/slapd/.libs/libslapd_la-getfilelist.o ldap/servers/slapd/.libs/libslapd_la-ldapi.o ldap/servers/slapd/.libs/libslapd_la-ldaputil.o ldap/servers/slapd/.libs/libslapd_la-lenstr.o ldap/servers/slapd/.libs/libslapd_la-libglobs.o ldap/servers/slapd/.libs/libslapd_la-localhost.o ldap/servers/slapd/.libs/libslapd_la-log.o ldap/servers/slapd/.libs/libslapd_la-mapping_tree.o ldap/servers/slapd/.libs/libslapd_la-match.o ldap/servers/slapd/.libs/libslapd_la-modify.o ldap/servers/slapd/.libs/libslapd_la-modrdn.o ldap/servers/slapd/.libs/libslapd_la-modutil.o ldap/servers/slapd/.libs/libslapd_la-object.o ldap/servers/slapd/.libs/libslapd_la-objset.o ldap/servers/slapd/.libs/libslapd_la-operation.o ldap/servers/slapd/.libs/libslapd_la-opshared.o ldap/servers/slapd/.libs/libslapd_la-pagedresults.o ldap/servers/slapd/.libs/libslapd_la-pblock.o ldap/servers/slapd/.libs/libslapd_la-plugin.o ldap/servers/slapd/.libs/libslapd_la-plugin_acl.o ldap/servers/slapd/.libs/libslapd_la-plugin_mmr.o ldap/servers/slapd/.libs/libslapd_la-plugin_internal_op.o ldap/servers/slapd/.libs/libslapd_la-plugin_mr.o ldap/servers/slapd/.libs/libslapd_la-plugin_role.o ldap/servers/slapd/.libs/libslapd_la-plugin_syntax.o ldap/servers/slapd/.libs/libslapd_la-protect_db.o ldap/servers/slapd/.libs/libslapd_la-proxyauth.o ldap/servers/slapd/.libs/libslapd_la-pw.o ldap/servers/slapd/.libs/libslapd_la-pw_retry.o ldap/servers/slapd/.libs/libslapd_la-rdn.o ldap/servers/slapd/.libs/libslapd_la-referral.o ldap/servers/slapd/.libs/libslapd_la-regex.o ldap/servers/slapd/.libs/libslapd_la-resourcelimit.o ldap/servers/slapd/.libs/libslapd_la-result.o ldap/servers/slapd/.libs/libslapd_la-rewriters.o ldap/servers/slapd/.libs/libslapd_la-sasl_map.o ldap/servers/slapd/.libs/libslapd_la-schema.o ldap/servers/slapd/.libs/libslapd_la-schemaparse.o ldap/servers/slapd/.libs/libslapd_la-security_wrappers.o ldap/servers/slapd/.libs/libslapd_la-slapd_plhash.o ldap/servers/slapd/.libs/libslapd_la-slapi_counter.o ldap/servers/slapd/.libs/libslapd_la-slapi2runtime.o ldap/servers/slapd/.libs/libslapd_la-snmp_collator.o ldap/servers/slapd/.libs/libslapd_la-sort.o ldap/servers/slapd/.libs/libslapd_la-ssl.o ldap/servers/slapd/.libs/libslapd_la-str2filter.o ldap/servers/slapd/.libs/libslapd_la-subentry.o ldap/servers/slapd/.libs/libslapd_la-task.o ldap/servers/slapd/.libs/libslapd_la-time.o ldap/servers/slapd/.libs/libslapd_la-thread_data.o ldap/servers/slapd/.libs/libslapd_la-uniqueid.o ldap/servers/slapd/.libs/libslapd_la-uniqueidgen.o ldap/servers/slapd/.libs/libslapd_la-upgrade.o ldap/servers/slapd/.libs/libslapd_la-utf8.o ldap/servers/slapd/.libs/libslapd_la-utf8compare.o ldap/servers/slapd/.libs/libslapd_la-util.o ldap/servers/slapd/.libs/libslapd_la-uuid.o ldap/servers/slapd/.libs/libslapd_la-value.o ldap/servers/slapd/.libs/libslapd_la-valueset.o ldap/servers/slapd/.libs/libslapd_la-vattr.o ldap/servers/slapd/.libs/libslapd_la-slapi_pal.o src/libsds/external/csiphash/.libs/libslapd_la-csiphash.o ldap/servers/slapd/.libs/libslapd_la-getsocketpeer.o ldap/libraries/libavl/.libs/libslapd_la-avl.o   -Wl,-rpath -Wl,/home/vashirov/src/ds/rpmbuild/BUILD/389-ds-base-2.1.0.20210913git9eaf1f35e/.libs -lssl -lcrypto -lldap_r -llber -lsasl2 -lkrb5 -lk5crypto -lcom_err -lpcre -lsystemd ./.libs/libsvrcore.so -lcrack -lc -lm -lrt -lutil -latomic -lssl3 -lsmime3 -lnss3 -lnssutil3 -lplds4 -lplc4 -lnspr4 -lpthread -ldl /home/vashirov/src/ds/rpmbuild/BUILD/389-ds-base-2.1.0.20210913git9eaf1f35e/rs/rslapd/release/librslapd.a -Wl,-z,now  -flto=thin -g -O2 -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -O2 -flto -g -grecord-gcc-switches -fstack-protector-strong -m64 -mtune=generic -Wl,-z -Wl,relro -Wl,--as-needed -Wl,-z -Wl,now -flto   -Wl,-soname -Wl,libslapd.so.0 -o .libs/libslapd.so.0.1.0
clang-12: warning: argument unused during compilation: '-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1' [-Wunused-command-line-argument]
/usr/include/c++/11/bits/stl_vector.h:1045: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = long unsigned int; _Alloc = std::allocator<long unsigned int>; std::vector<_Tp, _Alloc>::reference = long unsigned int&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__n < this->size()' failed.
clang-12: error: unable to execute command: Aborted (core dumped)
clang-12: error: linker command failed due to signal (use -v to see invocation)
make[2]: *** [Makefile:5776: libslapd.la] Error 254
make[2]: Leaving directory '/home/vashirov/src/ds/rpmbuild/BUILD/389-ds-base-2.1.0.20210913git9eaf1f35e'
make[1]: *** [Makefile:3993: all] Error 2
make[1]: Leaving directory '/home/vashirov/src/ds/rpmbuild/BUILD/389-ds-base-2.1.0.20210913git9eaf1f35e'
error: Bad exit status from /var/tmp/rpm-tmp.yf8mXz (%build)
```
Moreover, see failed checks. GCC doesn't support thin LTO, so pushing it to the default CFLAGS is not an option.
FTR, here are the default fedora LTO flags:
https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/rawhide/f/macros#_339-349

**Comment by vashirov** [comment]
I think this particular linker failure comes from `-Wp,-D_GLIBCXX_ASSERTIONS`
https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/rawhide/f/macros#_353

**Comment by vashirov** [comment]
Regarding linker's SIGABRT, turned out I had nightly rust toolchain enabled  :facepalm:, which uses LLVM 13. After switching to the default rustc from the repo, I managed to proceed further. 

Build with thin LTO still fails though. I disabled it in `Cargo.toml` and removed `-flto=thin` from `AM_CFLAGS` and `AM_CPPFLAGS` and the rpm build with clang finally succeeded. 

rpm build with clang and ASAN still fails, because `ld` is used instead of `lld`. I don't know why it is not used, even though `-fuse-ld=lld` is passed. I finally managed to build it after switch alternatives to use lld:
```
update-alternatives --set ld /usr/bin/ld.lld
```

**Comment by Firstyear** [comment]
Yeah I think that's a bug in autotools, that it's not correctly passing LDFLAGS into the C++ linking statements. I had to actualy set CXX to "clang++ -fuse=lld" to make it work before I built.

This is where I'm getting to, is that I think autotools is really limiting us at this point and what we are trying, and we're hitting a few issues all at once.

* gcc asan + rust asan (llvm) can't mix, or can't always mix in a stable way
* flto=thin relies on a consistent linker
* autotools has issues with LDFLAGS

We could try to use ld instead of lld with clang which may help here. But I think we'll just run into something else. 

I'll give it a try today and see. At some point I'm going to table flip and look at alternate build systems though.

**Comment by Firstyear** [comment]
Right, I tried clang++ with fuse=ld, and that's failed for a different set of reasons as well when we have LTO. So I think we're gonna need to focus on the profiles above and work out what to do from here. @mreynolds389 ping so you're aware. 

My immediate thought is that we fix "gcc + rust + ld + no-asan + lto=thin" for releases, because that's gonna affect fedora and suse together, but then after that we look into what we need to change in buildsystem inside autotools or swapping to something else to support the rest. It's gonna be a bit of pain shortterm, but it'll be better in the end since I think we're really hitting a lot of challenges and limits here.

**Comment by vashirov** [comment]
> Yeah I think that's a bug in autotools, that it's not correctly passing LDFLAGS into the C++ linking statements. I had to actualy set CXX to "clang++ -fuse=lld" to make it work before I built.

There are other things at play too: rpm build uses rpm macros that set additional CFLAGS/LDFLAGS and other linker/compiler parameters.
 
> * flto=thin relies on a consistent linker

Why do you insist on using **thin** LTO? Both Fedora and openSUSE don't use `-flto=thin`, just `-flto`.

> * clang + rust need to agree on their llvm version that's used

My mistake was to use nightly toolchain from rustup. If Rust comes from a distro, not from rustup, then it shouldn't be a problem:
```
$ ldd /usr/bin/rustc | grep -i LLVM
        libLLVM-12.so => /lib64/libLLVM-12.so (0x00007f3213f7c000)

$ rpm -qf /lib64/libLLVM-12.so
llvm-libs-12.0.1-1.fc34.x86_64
```

> I think we may end up needing to specify some limited use profiles. Maybe like:
> 
> * gcc + rust + ld + no-asan + lto=thin
> * clang + rust + lld + no-asan + lto=thin
> * clang + rust + lld + asan + lto=thin

Again, please see my question about thin LTO. I think just LTO would be enough for now. At least until we can iron out all build issues.
And I'm yet to see a problem with gcc+rust+asan (forgive me for being skeptical, it's part of the job :) ), so I'd like to enable it in PR CI, as right now it would give the biggest ROI.

**Comment by Firstyear** [comment]
> > Yeah I think that's a bug in autotools, that it's not correctly passing LDFLAGS into the C++ linking statements. I had to actualy set CXX to "clang++ -fuse=lld" to make it work before I built.
> 
> There are other things at play too: rpm build uses rpm macros that set additional CFLAGS/LDFLAGS and other linker/compiler parameters.

I'm not using rpm for these yet, just setting the flags manually, but yes I can see that being ... "fun" as well. 

> 
> > * flto=thin relies on a consistent linker
> 
> Why do you insist on using **thin** LTO? Both Fedora and openSUSE don't use `-flto=thin`, just `-flto`.

In gcc -flto is thin - in clang -flto is full lto, you need to specify -flto=thin. Effectively they are the same here though for all intents. It's just a gcc/clang-ism. 

But regardless, it's the same - you need everything to agree on either ld or lld. And it's clear that autotools won't use lld correctly with c++, and I'm seeing flto issues with ld + clang + rust, and ld + gcc + rust. Even trying -flto=true (full) it doesn't work. 

> 
> > I think we may end up needing to specify some limited use profiles. Maybe like:
> > 
> > * gcc + rust + ld + no-asan + lto=thin
> > * clang + rust + lld + no-asan + lto=thin
> > * clang + rust + lld + asan + lto=thin
> 
> Again, please see my question about thin LTO. I think just LTO would be enough for now. At least until we can iron out all build issues.
> And I'm yet to see a problem with gcc+rust+asan (forgive me for being skeptical, it's part of the job :) ), so I'd like to enable it in PR CI, as right now it would give the biggest ROI.

I've seen tons of issues with gcc + rust + asan, it's basicly been unusable for ages which is why I spent the time to dig into clang.  I will try to give an example of this today, but IMO it's at a point where we can't trust gcc's asan impl to mix with rust which is really disappointing. 

This is why I think we should focus on maybe two scenarios? 

* gcc + rust + ld + no-asan + lto=thin
* clang + rust + lld + asan + lto=thin

**Comment by Firstyear** [comment]
Okay, here's the issue with gcc + rust + asan:

```
   Compiling foreign-types-shared v0.1.1
     Running `sccache rustc --crate-name foreign_types_shared /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/foreign-types-shared-0.1.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 -C metadata=bb92834e32d6dca0 -C extra-filename=-bb92834e32d6dca0 --out-dir /home/william/build/ds/rs/pwdchan/debug/deps -L dependency=/home/william/build/ds/rs/pwdchan/debug/deps --cap-lints allow`
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: ./.libs/libslapd.so: undefined reference to `__asan_register_elf_globals'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: ./.libs/libslapd.so: undefined reference to `__asan_unregister_elf_globals'
collect2: error: ld returned 1 exit status
```

tl;dr gcc asan is missing some of the newer asan symbols, but it could just be due to the fact it's an older gcc in SLE compared to fedora/tw. Tried with gcc10 and it works, but I'm still skeptical. 

Anyway updated the PR and now the following work correctly

* gcc + asan + lto + rust
* clang + asan + rust + no-lto

**Comment by vashirov** [comment]
With this change I still have build failures in rpm build.

I had to make the following changes:
1. Disabled GCC security flags when building with clang.
2. lld by default uses xxhash for build ids, which is too small for rpm (it requires it between 16 and 64 bytes in size), use sha1 instead.

```diff
diff --git a/Makefile.am b/Makefile.am
index 9b644ea2e..0e4a5dde5 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -23,7 +23,6 @@ NQBUILDNUM := $(subst \,,$(subst $(QUOTE),,$(BUILDNUM)))
 DEBUG_DEFINES = @debug_defs@
 DEBUG_CFLAGS = @debug_cflags@
 DEBUG_CXXFLAGS = @debug_cxxflags@
-GCCSEC_CFLAGS = @gccsec_cflags@
 if CLANG_ENABLE
 ASAN_CFLAGS = @asan_cflags@
 else
@@ -32,6 +31,7 @@ ASAN_CFLAGS = @asan_cflags@ -lasan
 else
 ASAN_CFLAGS = @asan_cflags@
 endif
+GCCSEC_CFLAGS = @gccsec_cflags@
 endif
 MSAN_CFLAGS = @msan_cflags@
 TSAN_CFLAGS = @tsan_cflags@
@@ -88,7 +88,7 @@ endif
 
 if CLANG_ENABLE
 CLANG_ON = 1
-CLANG_LDFLAGS = -latomic -fuse-ld=lld
+CLANG_LDFLAGS = -latomic -fuse-ld=lld -Wl,--build-id=sha1
 EXPORT_LDFLAGS =
 else
 CLANG_ON = 0
```

3. DWARF5 is not supported yet by rpm in F34, `debugedit` fails to extract debuginfo and rpm build fails. 
@progier389, I see it was added as part of lmdb migration. Can we use DWARF4 until rpm start to support it (F35 I think)?
```diff
diff --git a/configure.ac b/configure.ac
index 51d762aa3..e82155d62 100644
--- a/configure.ac
+++ b/configure.ac
@@ -125,8 +125,8 @@ AC_ARG_ENABLE(debug, AS_HELP_STRING([--enable-debug], [Enable debug features (de
 AC_MSG_RESULT($enable_debug)
 if test "$enable_debug" = yes ; then
   debug_defs="-DDEBUG -DMCC_DEBUG"
-  debug_cflags="-g3 -ggdb -gdwarf-5  -O0"
-  debug_cxxflags="-g3 -ggdb -gdwarf-5  -O0"
+  debug_cflags="-g3 -ggdb -gdwarf-4 -O0"
+  debug_cxxflags="-g3 -ggdb -gdwarf-4 -O0"
   debug_rust_defs="-C debuginfo=2 -Z macro-backtrace"
   cargo_defs=""
   rust_target_dir="debug"
@@ -242,9 +242,6 @@ else
   # Without this, -fPIC doesn't work on generic fedora builds, --disable-gcc-sec.
   gccsec_cflags=""
 fi
-AM_COND_IF([RPM_HARDEND_CC],
-           [ gccsec_cflags="$gccsec_flags -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1" ],
-           [])
 AC_SUBST([gccsec_cflags])
 
 # Pull in profiling.
```

4. Disable LTO for clang rpm build:
```diff
diff --git a/rpm/389-ds-base.spec.in b/rpm/389-ds-base.spec.in
index 0700e5bc1..6afdd8f6c 100644
--- a/rpm/389-ds-base.spec.in
+++ b/rpm/389-ds-base.spec.in
@@ -37,7 +37,7 @@
 
 %if %{use_clang}
 %global toolchain clang
-%global _missing_build_ids_terminate_build 0
+%global _lto_cflags %nil
 %endif
 
 # Build cockpit plugin

```

These configurations were successfully built using rpm build:
* gcc + rust + lto (default on Fedora)
* gcc + rust + lto + asan
* clang + rust + w/o lto
* clang + rust + w/o lto + asan

**Comment by progier389** [comment]
> 1. DWARF5 is not supported yet by rpm in F34, `debugedit` fails to extract debuginfo and rpm build fails.
>    @progier389, I see it was added as part of lmdb migration. Can we use DWARF4 until rpm start to support it (F35 I think)?
> 

Of course ! no problem with that !

**Comment by Firstyear** [comment]
> These configurations were successfully built using rpm build:
> 
> * gcc + rust + lto (default on Fedora)
> * gcc + rust + lto + asan
> * clang + rust + w/o lto
> * clang + rust + w/o lto + asan

Cool, is there a PR for these changes so I can test it?

**Comment by Firstyear** [comment]
@vashirov ping re a PR with your changes that resolve this?

**Comment by vashirov** [comment]
@Firstyear, here you go: https://github.com/389ds/389-ds-base/pull/4929 
It's based on this PR and includes your commit.
